### ISSUE TITLE
RFE: Better offline/async IMAP support

### ISSUE THREAD

**Issue by nicowilliams** [neomutt/neomutt#970]
* Why do you not like the current state?

Opening IMAP mailboxes is slow, even with the header cache.  IMAP disconnects -> mailbox close.

    * What would you like to see to change?

More caching.  Async loading of mailbox changes on open.  On connection reset or close by server, reconnect automatically and transparently.

A lot like using offlineimap, but without using offlineimap.

**Comment by 4nd3r** [comment]
blue sky!

**Comment by nicowilliams** [comment]
:(

Depending on how well abstracted the internal mutt interface to mailboxes is, it might be easy, or it might require refactoring lots of code to make that interface well abstracted.

What I'm thinking I'd really like is a SQLite3 mailbox type, and then a pseudo-mailbox type that combines a SQLite3 mailbox with an IMAP mailbox, with the former acting as a cache for the latter, and with an option/variant to download and keep as much mail in the SQLite3 mbox as possible (a download from IMAP mode as opposed to cache IMAP mode).

**Comment by nicowilliams** [comment]
At a quick glance it looks like this would be a huge undertaking :(

The assumption that there's a list of all the messages in the folder is too baked in.

Instead, what I'd like is for each page of the index menu to be computed by querying the mailbox backend.  A lot like a SQL query with `ORDER BY` and `OFFSET`/`LIMIT` clauses.  Restructuring the code to do this seems like the first requirement.

**Comment by flatcap** [comment]
> A lot like using offlineimap, but without using offlineimap.

hehe, you don't ask for much :-)

It's a reasonable request, but we have quite a bit of work to get there.

Currently, NeoMut is single-threaded and consists of a large heap of structures that are badly intertwined with the user interface (ncurses) and a mass of global variables.

One of the sub-systems that we're currently trying to make independent is the socket layer.
When that's done we could conceivably make the IMAP code independent.
At that point we could have asynchronous IMAP connections.

We only have a few active developers, and we're working in our free time, so progress is slow.
But we **are** progressing.

We welcome volunteers.  We can always make time to help others, help us :-)

**Comment by nicowilliams** [comment]
Eh, I know.  I've looked at mutt source a number of times to think about how to implement this.  Each time I find it so daunting that I end up thinking it would be better to write a new MUA from scratch.

I don't think threads are needed here, provided there's an event loop that's entered often enough (there is, at the top of each menu).  Though I agree that threads would make some of this easier.

But even ignoring all the complexity of async or threaded IMAP, the need to load the mailbox, rather than ask it to paginate (and search), is architecturally rather broken.  Imagine opening and browsing a folder with millions of messages!

Incidentally, I rarely delete or file email.  I do use filters to route email to folders by mailing list, but even so, I accumulate huge numbers of emails in my folders, and startup time is very long.  I used to even not filter email _at all_, instead using `l` to filter my view -- that should scale, and it can scale, but not the way mutt works.
